equinix/resource_metal_virtual_circuit.go

```diff
@@ -146,141 +145,173 @@ func resourceMetalVirtualCircuit() *schema.Resource {
 	}
 
-	vc, _, err := client.VirtualCircuits.Create(projectId, connId, portId, &vncr, nil)
+	vc, _, err := client.InterconnectionsApi.CreateInterconnectionPortVirtualCircuit(ctx, connId, portId).VirtualCircuitCreateInput(vncr).Execute()
 	if err != nil {
 		log.Printf("[DEBUG] Error creating virtual circuit: %s", err)
 		return diag.FromErr(err)
 	}
+
+	var vcId string
+
+	if vc.VlanVirtualCircuit != nil {
+		vcId = vc.VlanVirtualCircuit.GetId()
+	} else {
+		vcId = vc.VrfVirtualCircuit.GetId()
+	}
+
 	// TODO: offer to wait while VCStatusPending
 	createWaiter := getVCStateWaiter(
+		ctx,
 		client,
-		vc.ID,
+		vcId,
 		d.Timeout(schema.TimeoutCreate)-30*time.Second,
-		[]string{string(packngo.VCStatusActivating)},
-		[]string{string(packngo.VCStatusActive)},
+		[]string{string(metalv1.VLANVIRTUALCIRCUITSTATUS_ACTIVATING)},
+		[]string{string(metalv1.VLANVIRTUALCIRCUITSTATUS_ACTIVE)},
 	)
 
 	_, err = createWaiter.WaitForStateContext(ctx)
 	if err != nil {
-		return diag.Errorf("Error waiting for virtual circuit %s to be created: %s", vc.ID, err.Error())
+		return diag.Errorf("Error waiting for virtual circuit %s to be created: %s", vcId, err.Error())
 	}
 
-	d.SetId(vc.ID)
+	d.SetId(vcId)
 
 	return resourceMetalVirtualCircuitRead(ctx, d, meta)
 }
 
 func resourceMetalVirtualCircuitRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	meta.(*config.Config).AddModuleToMetalUserAgent(d)
-	client := meta.(*config.Config).Metal
+	client := meta.(*config.Config).NewMetalClientForSDK(d)
 	vcId := d.Id()
 
-	vc, _, err := client.VirtualCircuits.Get(
-		vcId,
-		&packngo.GetOptions{Includes: []string{"project", "virtual_network", "vrf"}},
-	)
+	vc, _, err := client.InterconnectionsApi.GetVirtualCircuit(ctx, vcId).
+		Include([]string{"project", "virtual_network", "vrf"}).
+		Execute()
+
 	if err != nil {
 		return diag.FromErr(err)
 	}
 
 	// TODO: use API field from VC responses when available The regexp is
 	// optimistic, not guaranteed. This affects resource imports. "port" is not
 	// in the Includes above to assure the Href needed below.
+	var portHref string
+
+	if vc.VlanVirtualCircuit != nil {
+		portHref = vc.VlanVirtualCircuit.Port.GetHref()
+	} else {
+		portHref = vc.VlanVirtualCircuit.Port.GetHref()
+	}
 	connectionID := "" // vc.Connection.ID is not available yet
 	portID := ""       // vc.Port.ID would be available with ?include=port
 	connectionRe := regexp.MustCompile("/connections/([0-9a-z-]+)/ports/([0-9a-z-]+)")
-	matches := connectionRe.FindStringSubmatch(vc.Port.Href.Href)
+	matches := connectionRe.FindStringSubmatch(portHref)
 	if len(matches) == 3 {
 		connectionID = matches[1]
 		portID = matches[2]
 	} else {
-		log.Printf("[DEBUG] Could not parse connection and port ID from port href %s", vc.Port.Href.Href)
+		log.Printf("[DEBUG] Could not parse connection and port ID from port href %s", portHref)
 	}
+	errs := &multierror.Error{}
 
-	err = equinix_schema.SetMap(d, map[string]interface{}{
```

Comment:
The more granular type mapping means we can't just look up whatever fields we want and shove whatever we get into a map; we have to know if we're working with a VLAN or VRF virtual circuit and look up only the appropriate fields based on the type.

This could be seen as a shortcoming of strict schema compliance but it's also possible this means that there should be separate resources for `vlan_virtual_circuit` and `vrf_virtual_circuit`.